applications/OptimizationApplication/custom_utilities/geometrical/symmetry_utility.cpp

```diff
@@ -0,0 +1,270 @@
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "spatial_containers/spatial_containers.h"
+#include "utilities/builtin_timer.h"
+#include "utilities/parallel_utilities.h"
+#include "optimization_application.h"
+
+#define PI 3.14159265
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+SymmetryUtility::SymmetryUtility(std::string Name, ModelPart& rModelPart, Parameters SymmetrySettings )
+    : mUtilName(Name), mrModelPart( rModelPart ), mSymmetrySettings( SymmetrySettings )
+{
+
+    if(mSymmetrySettings["type"].GetString()=="plane_symmetry"){
+        mPlaneSymmetry = true;
+        mPlaneSymmetryData.Point = mSymmetrySettings["settings"]["point"].GetVector();
+        array_3d normal = mSymmetrySettings["settings"]["normal"].GetVector();
+        const double norm_normal = norm_2(normal);
+        if (norm_normal < std::numeric_limits<double>::epsilon())
+            KRATOS_ERROR<<"SymmetryUtility: norm of normal is close to zero"<<std::endl;
+            
+        mPlaneSymmetryData.Normal = normal/norm_normal;
+        mPlaneSymmetryData.ReflectionMatrix = IdentityMatrix(3) - (2*outer_prod(mPlaneSymmetryData.Normal, mPlaneSymmetryData.Normal));
+    }
+    else if(mSymmetrySettings["type"].GetString()=="rotational_symmetry"){
+        mAxisSymmetry = true;
+        mRotationalSymmetryData.Point = mSymmetrySettings["settings"]["point"].GetVector();
+        array_3d axis = mSymmetrySettings["settings"]["axis"].GetVector();
+        const double norm_axis = norm_2(axis);
+        if (norm_axis < std::numeric_limits<double>::epsilon())
+            KRATOS_ERROR<<"SymmetryUtility: norm of axis is close to zero"<<std::endl;
+            
+        mRotationalSymmetryData.Axis = axis/norm_axis;
+        mRotationalSymmetryData.Angle = mSymmetrySettings["settings"]["angle"].GetDouble();        
+    }
+    else
+        KRATOS_ERROR<<"SymmetryUtility: type should be either plane or axis"<<std::endl;
+
+}
+
+
+
+void SymmetryUtility::Initialize()
+{
+
+    BuiltinTimer timer;
+    KRATOS_INFO("SymmetryUtility::Initialize: ") << "Starting initialization of symmetry utility "<<mUtilName<<" ..." << std::endl;
+
+    NodeVector mListOfNodesInModelPart;
+    mListOfNodesInModelPart.resize(mrModelPart.Nodes().size());
+    int counter = 0;
+    for (ModelPart::NodesContainerType::iterator node_it = mrModelPart.NodesBegin(); node_it != mrModelPart.NodesEnd(); ++node_it)
+    {
+        NodeTypePointer pnode = *(node_it.base());
+        mListOfNodesInModelPart[counter++] = pnode;
+    }
+
+    const size_t bucket_size = 100;
+    KDTree search_tree(mListOfNodesInModelPart.begin(), mListOfNodesInModelPart.end(), bucket_size);
+
+    if(mAxisSymmetry){
+
+        #pragma omp declare reduction (merge : std::vector<std::pair <NodeTypePointer,NodeVector>> : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()))        
```

Comment:
I am aware of this util. Since it does not offer the used functionality and on the other hand performance on linux machine matters for now, I would keep this implementation.

Reply:
The `ParallelUtilities` with `ReductionUtilities` offers this functionality, hence I would use it. Following is an example:
https://github.com/KratosMultiphysics/Kratos/blob/fe0dacad112f45990ec362a3ed2c9c64a29fdbf9/applications/FreeSurfaceApplication/custom_utilities/edgebased_levelset_substep.h#L588-L595